pkg/querier/queryrange/downstreamer.go

```diff
@@ -0,0 +1,185 @@
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/loki/pkg/logql"
+)
+
+const (
+	DefaultDownstreamConcurrency = 32
+)
+
+type DownstreamHandler struct {
+	next queryrange.Handler
+}
+
+func ParamsToLokiRequest(params logql.Params) *LokiRequest {
+	return &LokiRequest{
+		Query:     params.Query(),
+		Limit:     params.Limit(),
+		Step:      int64(params.Step() / time.Millisecond),
+		StartTs:   params.Start(),
+		EndTs:     params.End(),
+		Direction: params.Direction(),
+		Path:      "/loki/api/v1/query_range", // TODO(owen-d): make this derivable
+	}
+}
+
+func (h DownstreamHandler) Downstreamer() logql.Downstreamer {
+	p := DefaultDownstreamConcurrency
+	locks := make(chan struct{}, p)
+	for i := 0; i < p; i++ {
+		locks <- struct{}{}
+	}
+	return &instance{
+		parallelism: p,
+		locks:       locks,
+		handler:     h.next,
+	}
+}
+
+// instance is an intermediate struct for controlling concurrency across a single query
+type instance struct {
+	parallelism int
+	locks       chan struct{}
+	handler     queryrange.Handler
+}
+
+func (i instance) Downstream(ctx context.Context, queries []logql.DownstreamQuery) ([]logql.Result, error) {
+	return i.For(queries, func(qry logql.DownstreamQuery) (logql.Result, error) {
+		req := ParamsToLokiRequest(qry.Params).WithShards(qry.Shards).WithQuery(qry.Expr.String()).(*LokiRequest)
+		logger, ctx := spanlogger.New(ctx, "DownstreamHandler.instance")
+		defer logger.Finish()
+		level.Debug(logger).Log("shards", req.Shards, "query", req.Query)
+
+		res, err := i.handler.Do(ctx, req)
+		if err != nil {
+			return logql.Result{}, err
+		}
+		return ResponseToResult(res)
+	})
+}
+
+// For runs a function against a list of queries, collecting the results or returning an error. The indices are preserved such that input[i] maps to output[i].
+func (in instance) For(
+	queries []logql.DownstreamQuery,
+	fn func(logql.DownstreamQuery) (logql.Result, error),
+) ([]logql.Result, error) {
+	type resp struct {
+		i   int
+		res logql.Result
+		err error
+	}
+
+	done := make(chan struct{})
+	defer close(done)
+
+	ch := make(chan resp)
+
+	// Make one goroutine to dispatch the other goroutines, bounded by instance parallelism
+	go func() {
+		for i := 0; i < len(queries); i++ {
+			select {
+			case <-done:
+				break
+			case <-in.locks:
+				go func(i int) {
+					// release lock back into pool
+					defer func() {
+						in.locks <- struct{}{}
+					}()
+
+					res, err := fn(queries[i])
+					response := resp{
+						i:   i,
+						res: res,
+						err: err,
+					}
+
+					// Feed the result into the channel unless the work has completed.
+					select {
+					case <-done:
+					case ch <- response:
+					}
+				}(i)
+			}
+		}
+	}()
+
+	results := make([]logql.Result, len(queries))
+	for i := 0; i < len(queries); i++ {
+		resp := <-ch
+		if resp.err != nil {
+			return nil, resp.err
+		}
+		results[resp.i] = resp.res
+	}
```

Comment:
A lot of code depends on the fact that len(results) == len(queries). I suggest you have a check here, in case it's not equal you could return an error.

This way we avoid users to panic on out of array bounds.

Reply:
Actually I didn't read the code but that for does exactly that.